File: iteration#1/src/main/java/model/StudentTranscript.java
package model;

import java.util.ArrayList;

public class StudentTranscript {
    private Student student;
    private ArrayList<Grade> letterGrades;

    public StudentTranscript(Student student){
        this.student = student;
        letterGrades = new ArrayList<Grade>();
    }

    public void passCourse(Course course, String letterGrade){
        letterGrades.add(new Grade(course, letterGrade));
    }

    public Integer completedCredits(){
        Integer answer = 0;
        for(Grade grade:letterGrades){
            if(grade.getLetterGrade().equals("FF")||grade.getLetterGrade().equals("FD")){continue;}
            //FF and FD aren't counted as completed.
            answer += grade.getCourseCredit();
        }
        return answer;
    }

    public Double getCGPA(){
        Integer sumOfCredits = 0;
        Double answer = 0.0;
        for(Grade grade:letterGrades) {
            sumOfCredits += grade.getCourseCredit();
            answer = answer + ((double)grade.getCourseCredit()) * grade.getCourseGPA();
        }

        if(sumOfCredits==0){return 0.0;}
        //This student is a new university student and hasn't taken any course yet. So her CGPA is assumed to be zero.


        answer /= sumOfCredits;
        return answer;
    }

    public String getCourseGrade(Course course){
        for(Grade grade:letterGrades){
            if(grade.getCourse().getCourseID().equals(course.getCourseID())){
                return grade.getLetterGrade();
            }
        }
        return null;
    }

    public ArrayList<Grade> getLetterGrades(){
        return letterGrades;
    }

    public void printStudentInfoLine(String line){
        for(int i=0;i<59;i++){
            if(i<line.length()){
                System.out.print(line.charAt(i));
            }
            else
                System.out.print(" ");
        }
        System.out.println("|");
    }
    public void printCourseInfoLine(String code,String courseName,String credit,String grade){
        String first = "| " + code;
        for(int i=0;i<10;i++){
            if(i<first.length()){
                System.out.print(first.charAt(i));
            }
            else{
                System.out.print(" ");
            }
        }
        String second = "| " + courseName;
        for(int i=0;i<26;i++){
            if(i<second.length()){
                System.out.print(second.charAt(i));
            }
            else{
                System.out.print(" ");
            }
        }
        String third = "| " + credit;
        for(int i=0;i<11;i++){
            if(i<third.length()){
                System.out.print(third.charAt(i));
            }
            else{
                System.out.print(" ");
            }
        }
        String fourth = "| " + grade;
        for(int i=0;i<12;i++){
            if(i<fourth.length()){
                System.out.print(fourth.charAt(i));
            }
            else{
                System.out.print(" ");
            }
        }
        System.out.println("|");
    }
    public void printStudentTranscript(){
        System.out.println("____________________________________________________________");
        System.out.println("|                                                          |");
        System.out.println("|                     TRANSCRIPT                           |");
        printStudentInfoLine("|  Student Name: "+student.getStudentName());
        printStudentInfoLine("|  Student ID: "+student.getStudentID());
        printStudentInfoLine("|  Student CGPA: "+student.getCGPA());
        printStudentInfoLine("|  Student Semester: "+student.getStudentSemester());
        System.out.println("____________________________________________________________");
        System.out.println("| Code    |   Course Name           | Credit   | Grade     |");
        System.out.println("____________________________________________________________");
        for(int i=0;i<letterGrades.size();i++){
            Course course = letterGrades.get(i).getCourse();
            printCourseInfoLine(course.getCourseID(),course.getCourseFullName(),
                    course.getCourseCredit().toString(),letterGrades.get(i).getLetterGrade());
        }
    }
}

File: Python/model/studentTranscript.py
from tkinter.messagebox import NO
from model.grade import Grade

class StudentTranscript:
  def __init__(self,student):
      self.student = student
      self.letter_grades = []

  def pass_course(self, course,letter_grade):
    self.letter_grades.append(Grade(course, letter_grade))

  def completed_credits(self):
    answer = 0
    for grade in self.letter_grades:
      if(grade.get_letter_grade() == "FF" or grade.get_letter_grade() == "FD"):
        continue
      answer += grade.get_course_credit()
    return answer

  def get_cgpa(self):
    sum_of_credits = 0
    answer = 0.0
    for grade in self.letter_grades:
      sum_of_credits += grade.get_course_credit()
      answer = answer + (float(grade.get_course_credit())) * grade.get_course_gpa()

    if(sum_of_credits==0):
      return 0.0
      # This student is a new university student and hasn't taken any course yet. So her CGPA is assumed to be zero.
    answer /= sum_of_credits
    return answer

  def get_course_grade(self, course):
      for grade in self.letter_grades:
        if(grade.get_course().get_course_id() == course.get_course_id()):
          return grade.get_letter_grade()
      return None

  def get_letter_grades(self):
    return self.letter_grades

  def print_student_info_line(self,line):
    for i in range(0, 59):
      if(i < len(line)):
        print(line[i])
      else:
        print(" ")
    print("|")

  def print_course_info_line(self, code, course_name, credit, grade):
    first = "| " + code
    for i in range(0, 10):
      if(i < len(first)):
        print(first[i])
      else:
        print(" ")

    second = "| " + course_name
    for i in range(0, 26):
      if(i < len(second)):
        print(second[i])
      else:
        print(" ")

    third = "| " + credit
    for i in range(0, 11):
      if(i < len(third)):
        print(third[i])
      else:
        print(" ")
    
    fourth = "| " + grade
    for i in range(0, 12):
      if(i < len(fourth)):
        print(fourth[i])
      else:
        print(" ")
    print("|")

  def print_student_transcript(self):
    print("____________________________________________________________");
    print("|                                                          |");
    print("|                     TRANSCRIPT                           |");
    self.print_student_info_line("|  Student Name: "+self.student.get_student_name())
    self.print_student_info_line("|  Student ID: "+self.student.get_student_id())
    self.print_student_info_line("|  Student CGPA: "+self.student.get_cgpa())
    self.print_student_info_line("|  Student Semester: "+self.student.getStudentSemester())
    print("____________________________________________________________")
    print("| Code    |   Course Name           | Credit   | Grade     |")
    print("____________________________________________________________")
    for i in range(len(self.letter_grades)):
            course = self.letter_grades[i].get_course();
            self.print_course_info_line(course.get_course_id(), course.get_course_full_name(), str(course.get_course_credit()), self.letter_grades[i].get_letter_grade())
